### PR TITLE
refs #8473 - API for retrieving environments and states

### PR DESCRIPTION
**Pull request by stbenjam**
Some refactoring, and adding support for retrieving environment and states list.

### CODE REVIEW

File: lib/smart_proxy_salt/rest.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+                                                       :client => 'runner')['return'][0]
+
+      files.each do |file|
+        if file =~ /\.sls\Z/
+          states << file.gsub('init.sls', '')
+                        .gsub('.sls', '')
+                        .chomp('/')
+                        .gsub('/', '.')
+        end
+      end
```
Comment (stbenjam) [inline lib/smart_proxy_salt/rest.rb]: Given the following directory structure:
```
.
├── apache
│   ├── bar.sls
│   ├── baz
│   │   ├── asdf.sls
│   │   └── qux.sls
│   ├── foo.sls
│   └── init.sls
└── core.sls
```

This returns the following list of states:
```
["apache.bar", "apache.baz.asdf", "apache.baz.qux", "apache.foo", "apache", "core"]
```
I probably should improve the processing a be a bit more resilient... but  @bechtoldt Does the result seem right?

File: lib/smart_proxy_salt/rest.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  class << self
+    def environments_list
+      Proxy::Salt::ApiRequest.new.post('/run', :fun => 'fileserver.envs', :client => 'runner')['return'][0].to_s
+    end
+
+    def states_list(environment)
+      states = []
+      files = Proxy::Salt::ApiRequest.new.post('/run', :fun => 'fileserver.file_list',
+                                                       :saltenv => environment,
+                                                       :client => 'runner')['return'][0]
```
Comment (stbenjam) [inline lib/smart_proxy_salt/rest.rb]: Returns:

```
["base", "dev"]
```

File: lib/smart_proxy_salt/rest.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
+require 'smart_proxy_salt/api_request'
+require 'json'
+
+module Proxy::Salt::Rest
+  extend ::Proxy::Log
+  extend ::Proxy::Util
+
+  class << self
+    def environments_list
+      Proxy::Salt::ApiRequest.new.post('/run', :fun => 'fileserver.envs', :client => 'runner')['return'][0].to_s
```
Comment (stbenjam) [inline lib/smart_proxy_salt/rest.rb]: Returns:

```
["base", "dev"]
```

**Comment by bechtoldt** [comment]
Looks promising. I'm going to setup a vagrant box with salt+foreman. Give me some days, I'll test these changes.

**Comment by stbenjam** [comment]
@bechtoldt That would be great!  It's still a very rough work in progress, but the foreman side is here https://github.com/theforeman/foreman_salt/pull/35.

**Pull request merged**